derive/src/lib.rs

```diff
@@ -16,110 +18,209 @@ mod items;
-	let source = input.to_string();
-	let ast = syn::parse_item(&source).expect("Failed to parse derive input");
+pub fn eth_abi(
+	args: proc_macro::TokenStream,
+	input: proc_macro::TokenStream
+)
+	-> proc_macro::TokenStream
+{
+    let args: proc_macro2::TokenStream = args.into();
 
 	let args = parse_args(args);
-	let endpoint_name = args.get(0).expect("Failed to parse an endpoint name argument");
-	let intf = items::Interface::from_item(ast);
-
-	match args.len() {
-		arg_count @ 1 | arg_count @ 2 => {
-			write_json_abi(&intf);
-			let name_ident_use: syn::Ident = format!("super::{}", &intf.name().clone()).into();
-			let mod_name = format!("pwasm_abi_impl_{}", &intf.name().clone());
-			let mod_name_ident: syn::Ident = mod_name.clone().into();
-			match arg_count {
-				1 => {
-					let endpoint = generate_eth_endpoint(&endpoint_name, &intf);
-					let endpoint_use: syn::Ident = format!("self::{}::{}", &mod_name, &endpoint_name).into();
-					let generated = quote! {
-						#intf
-						#[allow(non_snake_case)]
-						mod #mod_name_ident {
-							extern crate pwasm_ethereum;
-							extern crate pwasm_abi;
-							use pwasm_abi::types::*;
-							use #name_ident_use;
-							#endpoint
-						}
-						pub use #endpoint_use;
-					};
-					generated.parse().expect("Failed to parse generated input")
-				},
-				2 => {
-					let client_name = args.get(1).expect("Failed to parse an client name argument");
-					let endpoint = generate_eth_endpoint(&endpoint_name, &intf);
-					let client = generate_eth_client(client_name, &intf);
-					let endpoint_use: syn::Ident = format!("self::{}::{}", &mod_name, &endpoint_name).into();
-					let client_use: syn::Ident = format!("self::{}::{}", &mod_name, &client_name).into();
-					let generated = quote! {
-						#intf
-						#[allow(non_snake_case)]
-						mod #mod_name_ident {
-							extern crate pwasm_ethereum;
-							extern crate pwasm_abi;
-							use pwasm_abi::types::*;
-							use #name_ident_use;
-							#endpoint
-							#client
-						}
-						pub use #endpoint_use;
-						pub use #client_use;
-					};
-					generated.parse().expect("Failed to parse generated input")
-				},
-				_ => { unreachable!(); }
-			}
+	let intf = items::Interface::from_item(
+		parse_macro_input!(input as syn::Item));
+
+	write_json_abi(&intf);
+
+	let output: proc_macro2::TokenStream = match args.client_name() {
+		None => {
+			generate_eth_endpoint_wrapper(&intf, args.endpoint_name())
+		},
+		Some(client_name) => {
+			generate_eth_endpoint_and_client_wrapper(&intf, args.endpoint_name(), client_name)
 		}
-		len => {
-			panic!("eth_abi marco takes one or two comma-separated arguments, passed {}", len);
+	};
+
+    output.into()
+}
+
+/// Generates the eth abi code in case of a single provided endpoint.
+fn generate_eth_endpoint_wrapper(
+	intf: &items::Interface,
+	endpoint_name: &str
+)
+	-> proc_macro2::TokenStream
+{
+	// === REFACTORING TARGET ===
```

Comment:
what is this about?

Reply:
Would be nice to lay down such kinda of plan in to the comment.

Reply:
True! Will do that!

Reply:
Done